File: vector3/rust/src/main.rs
mod vector;
use vector::Vector;
use std::time::Instant;

const ITERATIONS: usize = 100_000_000;

fn benchmark<F: FnMut()>(label: &str, mut f: F) {
    let start = Instant::now();
    f();
    let duration = start.elapsed();
    println!("{}: {:>.10} seconds", label, duration.as_secs_f64());
}

fn main() {
    let v1 = Vector::new(1.0, 2.0, 3.0);
    let v2 = Vector::new(4.0, 5.0, 6.0);

    println!("Technology: Rust");
    benchmark("Addition", || {
        let mut _result;
        for _ in 0..ITERATIONS {
            _result = v1.add(v2);
        }
    });

    benchmark("Subtraction", || {
        let mut _result;
        for _ in 0..ITERATIONS {
            _result = v1.sub(v2);
        }
    });

    benchmark("Cross Product", || {
        let mut _result;
        for _ in 0..ITERATIONS {
            _result = v1.cross(v2);
        }
    });

    benchmark("Scalar Multiplication", || {
        let mut _result;
        for _ in 0..ITERATIONS {
            _result = v1.scale(2.5);
        }
    });

    benchmark("Normalization", || {
        let mut _result;
        for _ in 0..ITERATIONS {
            _result = v1.normalize();
        }
    });

    benchmark("Angle", || {
        for _ in 0..ITERATIONS {
            let _ = v1.angle(v2);
        }
    });
}

File: vector3/go/main.go
package main

import (
    "fmt"
    "time"
    "vector_perf/vector"
)

const iterations = 100000000

func benchmark(label string, fn func()) {
    start := time.Now()
    fn()
    duration := time.Since(start).Seconds()
    fmt.Printf("%s: %.6f seconds\n", label, duration)
}

func main() {
    v1 := vector.New(1.0, 2.0, 3.0)
    v2 := vector.New(4.0, 5.0, 6.0)

    fmt.Printf("Technology: Go\n")
    benchmark("Addition", func() {
        for i := 0; i < iterations; i++ {
            _ = v1.Add(v2)
        }
    })

    benchmark("Subtraction", func() {
        for i := 0; i < iterations; i++ {
            _ = v1.Sub(v2)
        }
    })

    benchmark("Cross Product", func() {
        for i := 0; i < iterations; i++ {
            _ = v1.Cross(v2)
        }
    })

    benchmark("Scalar Multiplication", func() {
        for i := 0; i < iterations; i++ {
            _ = v1.Scale(2.5)
        }
    })

    benchmark("Normalization", func() {
        for i := 0; i < iterations; i++ {
            _ = v1.Normalize()
        }
    })

    benchmark("Angle", func() {
        for i := 0; i < iterations; i++ {
            _ = v1.Angle(v2)
        }
    })
}

